qutebrowser/browser/history.py

```diff
@@ -332,11 +350,22 @@ def add_url(self, url, title="", *, redirect=False, atime=None):
             if redirect or self._is_excluded(url):
                 return
 
-            self.completion.insert({
-                'url': self._format_completion_url(url),
+            f_url = self._format_completion_url(url)
+            result = self.completion.insert({
+                'url': f_url,
                 'title': title,
-                'last_atime': atime
-            }, replace=True)
+                'last_atime': atime,
+                'visits': 1,
+                'frecency': atime,
+            }, ignore=True)
+
+            if not result.rows_affected():
+                update = {'visits': 'visits + 1',
+                          'frecency': '{} + visits * {}'.format(
+                              atime, self.completion.FRECENCY_BONUS),
+                          'last_atime': atime}
+
+                self.completion.update(update, {'url': f_url}, escape=False)
```

Comment:
Would an [`upsert`](https://www.sqlite.org/draft/lang_UPSERT.html) be more appropriate?

Reply:
It would be great but the SQLite version required by Qt it's not there yet. UPSERT was implemented in SQLite 3.24.0 and in Debian, for example, libqt5sql5-sqlite only requires libsqlite3-0 (>= 3.7.3).

Reply:
Well, I'm not totally sure now, because Debian doesn't really package anything older than 3.27.2, and Qt docs indicate that they include 3.24.0 in Qt5.9 and 3.31.1 in >=Qt5.12. We still need to check the other platforms for which qutebrowser gets packaged.

https://doc.qt.io/qt-5.9/qtsql-attribution-sqlite.html
https://doc.qt.io/qt-5.12/qtsql-attribution-sqlite.html
https://doc.qt.io/qt-5.14/qtsql-attribution-sqlite.html
https://doc.qt.io/qt-5/qtsql-attribution-sqlite.html

Reply:
I checked and it seems we're safe, so I updated the code.
```
Ubuntu 16.04 xenial/main sqlite3 3.11.0
Debian Stable stable/main sqlite3 3.27.2
Debian Testing testing/main sqlite3 3.31.1
Debian Unstable main sqlite3 3.32.1
Fedora 26 release sqlite 3.19.1
Fedora 27 release sqlite 3.20.1
Fedora 28 release sqlite 3.22.0
Fedora 29 release sqlite 3.24.0
Fedora 30 release sqlite 3.26.0
Fedora 31 release sqlite 3.29.0
Fedora 32 release sqlite 3.31.1
Arch core sqlite 3.32.1
Gentoo dev-db/sqlite 3.31.1
Void Linux x86_64 x86_64 sqlite 3.32.1
nixpkgs stable sqlite 3.31.1
nixpkgs stable sqlite 3.30.1
nixpkgs stable sqlite 3.27.2+replication3
nixpkgs unstable sqlite 3.31.1
nixpkgs unstable sqlite 3.27.2+replication3
openSUSE Leap 15.0 main/oss sqlite3 3.23.1
openSUSE Leap 15.1 main/oss sqlite3 3.28.0
openSUSE Leap 15.2 main/oss sqlite3 3.28.0
openSUSE Tumbleweed sqlite3 3.31.1
Slackware 14.2 slackware sqlite 3.13.0
Slackware current slackware sqlite 3.32.1
Slackware64 14.2 slackware64 sqlite 3.13.0
Slackware64 current slackware64 sqlite 3.32.1
FreeBSD Ports databases/sqlite3 3.31.1
Chocolatey SQLite 3.32.1
Homebrew sqlite 3.32.1
Qt-5.9 sqlite 3.24.0
Qt-5.12 sqlite 3.31.1
Qt-5.14 sqlite 3.31.1
Qt-5.15 sqlite 3.31.1
```